Onboarding TK Django Exercise

danielsto opened the pull request:
# My solution the Recipe API for the TK Django Training Exercise.

I think all of the requested features are implemented, but I'm having a weird problem with the `GET /api/recipes/` which passes the tests, but sometimes doesn't show the list of recipes in Postman. Response data in the tests is the expected.

israel-tk commented on app/core/models.py:
```diff
@@ -0,0 +1,39 @@
+    """
+    name = models.CharField(max_length=255)
+    description = models.TextField()
+    ingredients = ArrayField(JSONField(), null=True)
```
I think you got it wrong. If the ingredients are already linked to recipes via a foreign key we don't need this array field here. You can delete this field in the model.

israel-tk commented on app/app/settings.py:
```diff
@@ -0,0 +1,114 @@
+
+DATABASES = {
+    'default': {
+        'ENGINE': 'django.db.backends.postgresql',
```
It is OK, but take into account projects usually make use of options where all the connection parameters are in one environment variable, which is usually called "database url". See https://github.com/joke2k/django-environ for example.

israel-tk commented on app/core/models.py:
```diff
@@ -0,0 +1,39 @@
+        if self.ingredients:
+            for ing in self.ingredients:
+                if not(isinstance(ing, dict)):
+                    ing = literal_eval(ing)
```
This is not a good idea. You should have custom code in the view where the recipes are created or updated to create/update the ingredients related to a recipe, but not here. Any `*_eval` option is a bad smell if the input is coming from the user.

Do something like https://www.django-rest-framework.org/api-guide/serializers/#writable-nested-representations instead.

israel-tk commented on app/recipe/serializers.py:
```diff
@@ -0,0 +1,35 @@
+        read_only_fields = ('id',)
+
+
+class IngredientListSerializer(serializers.ModelSerializer):
```
If you follow https://www.django-rest-framework.org/api-guide/serializers/#writable-nested-representations you will only need one serializer.

israel-tk commented on app/recipe/views.py:
```diff
@@ -0,0 +1,44 @@
+from recipe import serializers
+
+
+class IngredientViewSet(viewsets.ModelViewSet,
```
This viewset is not expected in the API. What we expect is one endpoint that returns the whole recipe with all its ingredients.

Example:
```
{
	“id”: 1,
	“name”: “Pizza”
	“description”: “Put it in the oven”,
	“ingredients”: [{“name”: “dough”}, {“name”: “cheese”}, {“name”: “tomato”}]
}
```

There is no `/ingredients/` endpoint expected.

israel-tk reviewed (changes requested):
1. The approach you used to deal with the ingredients with the recipe is not OK: it repeats data and makes the whole process cumbersome even with custom Python evals. Use nested serializers instead and remove the `Recipe.ingredients` array field. I can provide pointers to you if you want.
2. The `/ingredients` set of endpoints is not needed. We expect to get and manipulate all the data from the `/recipes` one.

danielsto commented:
Hi @israel-tk, I'm working on this review in-betwen onboarding meetings this week and it seems like I've been hitting a wall for a while. Maybe you can point me in the right direction.

- I deleted the `ArrayField` of `JSONField` from the `Recipe` model.
- Added a custom function inside the `RecipeSerializer` to create `Ingredient` objects as I create the `Recipe`.
- Haven't changed anything in the views yet.

I ran the tests again just to see how that was going and they are throwing 400 in the post requests and `TypeError: Direct assignment to the reverse side of a related set is prohibited. Use ingredients.set() instead` when I pass ingredients to a recipe via `Recipe.objects.create(name="Some", description="...", ingredients=[...])`. However, when I test the API from Postman, the browser or the CLI via curl it seems to work. Any ideas?

israel-tk commented:
Hey! Can we meet at 11 am today to look into this?

On Wed, Oct 7, 2020 at 2:57 PM Daniel <notifications@github.com> wrote:

> Hi @israel-tk <https://github.com/israel-tk>, I'm working on this review
> in-betwen onboarding meetings this week and it seems like I've been hitting
> a wall for a while. Maybe you can point me in the right direction.
>
>    - I deleted the ArrayField of JSONField from the Recipe model.
>    - Added a custom function inside the RecipeSerializer to create
>    Ingredient objects as I create the Recipe.
>    - Haven't changed anything in the views yet.
>
> I ran the tests again just to see how that was going and they are throwing
> 400 in the post requests and TypeError: Direct assignment to the reverse
> side of a related set is prohibited. Use ingredients.set() instead when I
> pass ingredients to a recipe via Recipe.objects.create(name="Some",
> description="...", ingredients=[...]). However, when I test the API from
> Postman, the browser or the CLI via curl it seems to work. Any ideas?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/danielsto/tk-django-exercise/pull/1#issuecomment-704916809>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFJFI6X3VHWPHRHHI2QFUWTSJRQS3ANCNFSM4SBRCXRQ>
> .
>


-- 

*Israel Saeta Pérez*

Software Developer @ TravelPerk <http://www.travelperk.com>

danielsto commented:
Hi @israel-tk, I think I solved the issues we talked about in the last two commits.

To solve the list view not rendering everything that was on the database without restarting the server I did the following:

```python
# app/recipe/views.py ~RecipeViewSet
 def get_queryset(self):
        substring = self.request.query_params.get('name')
        queryset = self.queryset
        ...
```
changed to 
```python
# app/recipe/views.py ~RecipeViewSet
 def get_queryset(self):
        substring = self.request.query_params.get('name')
        queryset = Recipe.objects.all()
        ...
```

israel-tk commented on app/recipe/serializers.py:
```diff
@@ -0,0 +1,56 @@
+from core.models import Ingredient, Recipe
+
+
+class IngredientSerializer(serializers.ModelSerializer):
```
you don't need this serializer, it is a leftover

israel-tk commented on app/recipe/views.py:
```diff
@@ -0,0 +1,35 @@
+        Partial update of a recipe and its related ingredients (patch)
+        """
+        recipe_object = self.get_object()
+        recipe_object.ingredients.all().delete()
```
If you already do this in the serializer I think you don't need to do it here as well.

israel-tk commented on app/recipe/serializers.py:
```diff
@@ -0,0 +1,56 @@
+
+        ingredient_list = validated_data.pop('ingredients', [])
+
+        if ingredient_list:
```
Take into account that this won't cover the case of setting the list of ingredients to an empty array. But it is minor.

israel-tk reviewed (approved):
Just minor comments.

israel-tk commented:
> Hi @israel-tk, I think I solved the issues we talked about in the last two commits.
> 
> To solve the list view not rendering everything that was on the database without restarting the server I did the following:
> 
> ```python
> # app/recipe/views.py ~RecipeViewSet
>  def get_queryset(self):
>         substring = self.request.query_params.get('name')
>         queryset = self.queryset
>         ...
> ```
> 
> changed to
> 
> ```python
> # app/recipe/views.py ~RecipeViewSet
>  def get_queryset(self):
>         substring = self.request.query_params.get('name')
>         queryset = Recipe.objects.all()
>         ...
> ```

Interesting, I guess for some reason `self.queryset` was evaluated and then stored in the class forever until server restart, that's why you got stale results.
